### ISSUE TITLE
virtctl: Native ssh client does not use supplied identity file

### ISSUE THREAD

**Issue by 0xFelix** [kubevirt/kubevirt#7072]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
/kind bug
> /kind enhancement

**What happened**:

I wanted to use virtctl to ssh into a VM. I specified a private key with `--identity-file`.

The identity file is not used and authentication fails.

`ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain`

**What you expected to happen**:

The identity file is used and authentication succeeds.

**How to reproduce it (as minimally and precisely as possible)**:

Run `virtctl ssh user@vm --identity-file=/path/to/privatekey`

**Anything else we need to know?**:

I tried it with an ed25519 and a rsa private key separately.

With the local ssh client both private keys can be used to successfully authenticate.

**Environment**:
- KubeVirt version (use `virtctl version`):
`Client Version: version.Info{GitVersion:"v0.49.0", GitCommit:"f8160a0b5eac2acafab74fd7916bfcaf0c39a6a3", GitTreeState:"clean", BuildDate:"2022-01-11T17:37:26Z", GoVersion:"go1.16.6", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{GitVersion:"v0.44.3-26-g08a0ea47c", GitCommit:"08a0ea47c8e883629fc6241116c1a4db4284212a", GitTreeState:"clean", BuildDate:"2021-11-26T10:05:29Z", GoVersion:"go1.16.6", Compiler:"gc", Platform:"linux/amd64"}`
- Kubernetes version (use `kubectl version`):
- VM or VMI specifications:
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release): Fedora 35 / RHEL 8.5
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

**Comment by dominikholler** [comment]
CC @rmohr

**Comment by rmohr** [comment]
/triage accepted

Looks like the commandline is not picked up in the case of native ssh.

**Comment by rmohr** [comment]
> * OS (e.g. from /etc/os-release): Fedora 35 / RHEL 8.5

@0xFelix I just see now that you use FC35, I think you are running into #6851. What is basically happening is, that you have the ssh agent running. `virtctl` is fetching all default keys from the agend and all fail to authenticate due to the crypto bug mentioned there. Even if you would specify the EC key on the commandline it would not necessarily ever be used because the ssh server responds with a too-many-authentication failures before the EC key would be tried.

To confirm that this is the issue, use an EC kay and prefix `virtctl` like this to not use the ssh agent:

```
SSH_AUTH_SOCK=/dev/null virtctl ssh
```

**Comment by 0xFelix** [comment]
@rmohr I can confirm that authentication fails with `ssh: handshake failed: ssh: disconnect, reason 2: Too many authentication failures` when keys are loaded in my agent. But this was not my problem cause during my inital tests no keys were loaded. I think I must have missed the user or used the wrong key... it is working now.

**Comment by 0xFelix** [comment]
#7071 still persists though.

**Comment by rmohr** [comment]
> @rmohr #7071 still persists though.

:+1: 

And thanks for the updates.

**Comment by ordovicia** [comment]
Hi,
I succeeded to reproduce this issue.
It seems that when SSH agent is available, `virtctl` ignores `--identity-file` option, even when a key is not registered with the SSH agent.

### SSH agent is not available, key is specified => Success

```
$ SSH_AUTH_SOCK=/dev/null ./virtctl ssh ubuntu@test -i id_ed25519_kubevirt-test
# OK
```

### SSH agent is available, key is not registered but specified => Fail

```
$ printenv SSH_AUTH_SOCK
/home/my-name/.ssh/agent

$ ssh-add -T id_ed25519_kubevirt-test.pub
Agent signature failed for id_ed25519_kubevirt-test.pub: agent refused operation
# Key is not registered
```

```
$ ./virtctl ssh ubuntu@test -i id_ed25519_kubevirt-test
ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
```

### SSH agent is available, key is registered => Success

```
$ ssh-add -T id_ed25519_kubevirt-test.pub
# OK
```

```
$ ./virtctl ssh ubuntu@test
# OK
```

```
$ ./virtctl ssh ubuntu@test -i id_ed25519_kubevirt-test
# OK
```

### `virtctl version`

```
Client Version: version.Info{GitVersion:"v1.2.2", GitCommit:"1466b658f78b9b8bb9517ffb6dafd4b777f33fe6", GitTreeState:"clean", BuildDate:"2024-06-06T04:42:11Z", GoVersion:"go1.21.8 X:nocoverageredesign", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{GitVersion:"v1.2.2", GitCommit:"1466b658f78b9b8bb9517ffb6dafd4b777f33fe6", GitTreeState:"clean", BuildDate:"2024-06-06T05:46:26Z", GoVersion:"go1.21.8 X:nocoverageredesign", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by 0xFelix** [comment]
I think it comes down to the original issue where virtctl tries all other identities in the agent first and at this point retries are exhausted so it never tries the file you specified.

**Comment by ordovicia** [comment]
Ah, understood! Thank you for the explanation.
So, to prevent the issue, should `virtctl` try a specified identity file first before trying identities in SSH agent?
I guess the following two lines should be swapped?
https://github.com/kubevirt/kubevirt/blob/a1e9279695ff4508b3af9f3a814bfddab2670196/pkg/virtctl/ssh/native.go#L97-L98

**Comment by 0xFelix** [comment]
That's a good question. Could you try what happens when swapping those lines? I'm not sure what will happen if your key is protected by a password and you added it to the agent prior. We still want it to be picked up from the agent instead of virtctl asking for the password again.

**Comment by ordovicia** [comment]
I patched `virtctl` to swap those lines and did some experiments.
Obtained results:

- I was able to make `virtctl` try a specified SSH private key first.
- When I registered an SSH private key with the SSH agent and I specified the key via `--identity` option, I needed to type the key's passphrase.

If you specify an SSH private key, you will need to type its passphrase even when the key is registered with the SSH agent.
However, this is not a problem because you can just skip specifying the key and you will be able to login.

Experiment details:

### Setup

I used `virtctl` based on v1.3.0 with a patch to swap those lines.

```
$ git status
HEAD detached at v1.3.0
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   pkg/virtctl/ssh/native.go

no changes added to commit (use "git add" and/or "git commit -a")
```

```diff
diff --git a/pkg/virtctl/ssh/native.go b/pkg/virtctl/ssh/native.go
index 4cae1e788b..758c788813 100644
--- a/pkg/virtctl/ssh/native.go
+++ b/pkg/virtctl/ssh/native.go
@@ -94,8 +94,8 @@ func (o *NativeSSHConnection) PrepareSSHClient(kind, namespace, name string) (*s
 func (o *NativeSSHConnection) getAuthMethods(kind, namespace, name string) []ssh.AuthMethod {
 	var methods []ssh.AuthMethod

-	methods = o.trySSHAgent(methods)
 	methods = o.tryPrivateKey(methods)
+	methods = o.trySSHAgent(methods)

 	methods = append(methods, ssh.PasswordCallback(func() (secret string, err error) {
 		password, err := readPassword(fmt.Sprintf("%s@%s/%s.%s's password: ", o.Options.SSHUsername, kind, name, namespace))
```

I generated an SSH key pair with a passphrase.
I created a VM and registered the SSH public key with the VM's authorized keys.

### Case: SSH agent is available, key is not registered but specified => Success

When I specified the SSH private key to the patched `virtctl`, I successfully logged in to the VM.

```
$ ./virtctl ssh ubuntu@test -i id_ed25519_kubevirt-test
Key id_ed25519_kubevirt-test requires a password:
Welcome to Ubuntu 22.04.4 LTS (GNU/Linux 5.15.0-116-generic x86_64)
...
```

### Case: Key is registered with the SSH agent

I registered the SSH private key with the agent.

```
$ ssh-add id_ed25519_kubevirt-test
Enter passphrase for id_ed25519_kubevirt-test:
Identity added: id_ed25519_kubevirt-test
```

I was able to login to the VM without specifying the key and without typing the passphrase.

```
$ ./virtctl ssh ubuntu@test
Welcome to Ubuntu 22.04.4 LTS (GNU/Linux 5.15.0-116-generic x86_64)
...
```

When I specified the key to the patched `virtctl`, I needed to input the passphrase.

```
$ ./virtctl ssh ubuntu@test -i id_ed25519_kubevirt-test
Key id_ed25519_kubevirt-test requires a password:
Welcome to Ubuntu 22.04.4 LTS (GNU/Linux 5.15.0-116-generic x86_64)
...
```

**Comment by rmohr** [comment]
Are you still seeing this on latest `virtctl` versions? I think the issue was that golang ssh back then did not fully support looking up offered authentication algorithms by the remote server and just tried one after the other, eventually exhausting all connection attempts. Does this still happen with latest `virtctl`?

**Comment by ordovicia** [comment]
Yes, the problem persists for `v1.3.0` (latest released version) and current `main` (adb4f2cb5e020678388848fd8283098adaaad889).
